File: DP/1463_1로 만들기/1463_1로 만들기.py
# N을 입력받으세요
N = int(input())

# dp를 수행할 배열을 만드세요
# dp[0] = 0 (사용 안함)
# dp[1] = 0 : 초기값
dp = [0 for _ in range(N+1)]

# dp를 수행하세요
for i in range(2, N+1):
    dp[i] = dp[i-1] + 1
    if (i%3 == 0):
        dp[i] = min(dp[i], dp[i//3]+1)
    if (i%2 == 0):
        dp[i] = min(dp[i], dp[i//2]+1)

# 연산을 사용하는 횟수의 최솟값을 출력하세요
print(dp[-1])

File: DP/1463_1로 만들기/1463_1로 만들기.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.util.Arrays;

// buttom-up DP
public class Main {
  public static void main(String[] args) throws IOException{
    // set needed components
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    int N = Integer.parseInt(br.readLine());
    int[] memo = new int[N+1];
    memo[0] = 0; memo[1] = 0;

    // do process
    for (int i = 2; i <= N; i++){
      memo[i] = memo[i-1] + 1;
      if (i % 3 == 0) memo[i] = Math.min(memo[i], memo[i/3] + 1);
      if (i % 2 == 0) memo[i] = Math.min(memo[i], memo[i/2] + 1);
    }

    // print answer and end program
    System.out.print(memo[N]);
    br.close();
    return;
  }
}
